src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
+
+        /// <summary>Base type for all generator implementations that plug into the base Random.</summary>
+        private abstract class ImplBase
         {
-            if (minValue > maxValue)
-            {
-                throw new ArgumentOutOfRangeException(nameof(minValue), SR.Format(SR.Argument_MinMaxValue, nameof(minValue), nameof(maxValue)));
-            }
+            public abstract double Sample();
 
-            long range = (long)maxValue - minValue;
-            return range <= int.MaxValue ?
-                (int)(Sample() * range) + minValue :
-                (int)((long)(GetSampleForLargeRange() * range) + minValue);
-        }
+            public abstract int Next();
 
-        public virtual double NextDouble() => Sample();
+            public abstract int Next(int maxValue);
 
-        public virtual void NextBytes(byte[] buffer)
+            public abstract int Next(int minValue, int maxValue);
+
+            public abstract long NextInt64();
+
+            public abstract long NextInt64(long maxValue);
+
+            public abstract long NextInt64(long minValue, long maxValue);
+
+            public abstract double NextDouble();
+
+            public abstract float NextSingle();
+
+            public abstract void NextBytes(byte[] buffer);
+
+            public abstract void NextBytes(Span<byte> buffer);
+        }
+
+        #region Implementations
+        /// <summary>
+        /// Provides an implementation of the xoshiro256** algorithm. This implementation is used
+        /// when no seed is specified and an instance of the base Randomclass is constructed.  As
+        /// such, we are free to implement however we see fit, withoutback compat concerns around
+        /// the sequence of numbers generated or what methods call what other methods.
+        /// </summary>
+        private sealed class Xoshiro256StarStarImpl : ImplBase
         {
-            if (buffer is null)
+            // NextUInt64 is based on the algorithm from http://prng.di.unimi.it/xoshiro256starstar.c:
+            //
+            //     Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)
+            //
+            //     To the extent possible under law, the author has dedicated all copyright
+            //     and related and neighboring rights to this software to the public domain
+            //     worldwide. This software is distributed without any warranty.
+            //
+            //     See <http://creativecommons.org/publicdomain/zero/1.0/>.
+
+            private ulong _s0, _s1, _s2, _s3;
+
+            public unsafe Xoshiro256StarStarImpl()
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.buffer);
+                do
+                {
+                    Guid g;
+
+                    // Seed the four ulongs (4 64-bit values) from two Guids (2 128-bit values)
+                    g = Guid.NewGuid();
```

Comment:
I think this PR should include a test that uses reflection to set the four ulongs to something, and then compares the next `n` results to some canned values from a reference implementation of this algorithm. That would ensure neither you nor subsequent maintainers have a ytpo. After all, the original implementation apparently did.

Reply:
Ideally it would call each of the methods a few times, not just fill a buffer with NextBytes.

Reply:
Also I think it should ideally include tables for the ranged overloads, and the legacy impl needs one for NextInt64 of course.